format-avro/src/main/java/io/cdap/plugin/format/avro/AvroToStructuredTransformer.java

```diff
@@ -91,9 +95,76 @@ public Schema convertSchema(org.apache.avro.Schema schema) throws IOException {
     if (schemaCache.containsKey(hashCode)) {
       structuredSchema = schemaCache.get(hashCode);
     } else {
-      structuredSchema = Schema.parseJson(schema.toString());
+      String strSchema = schema.toString();
+      String jsonSchema = preprocessSchema(new Gson().fromJson(strSchema, JsonObject.class)).toString();
+      structuredSchema = Schema.parseJson(jsonSchema);
       schemaCache.put(hashCode, structuredSchema);
     }
     return structuredSchema;
   }
+
+  private JsonObject preprocessSchema(JsonObject schema) {
+
+    if (!schema.has("type")) {
+      return schema;
+    }
+
+    JsonElement type = schema.get("type");
+
+    if (type.isJsonArray()) {   // Union
+      for (JsonElement subtype : type.getAsJsonArray()) {
+        if (!subtype.isJsonPrimitive()) {
+          preprocessSchema(subtype.getAsJsonObject());
+        } else if (customTypes.containsKey(subtype.getAsString())) {
+          schema.remove("type");
+          schema.add("type", customTypes.get(subtype.getAsString()));
+        }
+      }
+    } else if (type.isJsonObject()) {   // Unnamed Complex type
+      preprocessSchema(type.getAsJsonObject());
+    } else {
+      String typeName = type.getAsString();
+      switch (typeName) {
+        case "record":
+          if (!schema.has("fields")) {
+            return schema;
```

Comment:
Do you still need to put it into customTypes?

Reply:
If a record does not have the `fields` attribute, it's an error. I thought I'd immediately return it, but now I intend to just not handle this case explicitly so that an exception will be thrown in such a case.